src/mapgen/mapgen.cpp

```diff
@@ -1081,27 +1081,68 @@ void MapgenParams::writeParams(Settings *settings) const
 		bparams->writeParams(settings);
 }
 
-// Calculate edges of outermost generated mapchunks (less than
-// 'mapgen_limit'), and corresponding exact limits for SAO entities.
-void MapgenParams::calcMapgenEdges()
+
+void MapgenParams::calcMapgenEdges(s16 mg_limit, s16 csize_n, s16 ccmin, s16 ccmax,
+	s16 ccfmin, s16 ccfmax, s16 *mapgen_edge_min, s16 *mapgen_edge_max)
 {
-	if (m_mapgen_edges_calculated)
-		return;
+	// Effective mapgen limit, in blocks
+	// Uses same calculation as ServerMap::blockpos_over_mapgen_limit(v3s16 p)
+	s16 mapgen_limit_b = mg_limit / MAP_BLOCKSIZE;
+	// Effective mapgen limits, in nodes
+	s16 mapgen_limit_min = -mapgen_limit_b * MAP_BLOCKSIZE;
+	s16 mapgen_limit_max = (mapgen_limit_b + 1) * MAP_BLOCKSIZE - 1;
+	// Number of complete chunks from central chunk fullminp/fullmaxp
+	// to effective mapgen limits.
+	s16 numcmin = MYMAX((ccfmin - mapgen_limit_min) / csize_n, 0);
+	s16 numcmax = MYMAX((mapgen_limit_max - ccfmax) / csize_n, 0);
+	// Mapgen edges, in nodes
+	*mapgen_edge_min = ccmin - numcmin * csize_n;
+	*mapgen_edge_max = ccmax + numcmax * csize_n;
+}
+
 
+// Calculate edges of outermost generated mapchunks.
+// Calculate exact limits for SAO entities.
+// Calculate spawn range maximum.
+
+void MapgenParams::calcMgEdgesAndSpawnLimit()
+{
 	// Central chunk offset, in blocks
 	s16 ccoff_b = -chunksize / 2;
 	// Chunksize, in nodes
-	s32 csize_n = chunksize * MAP_BLOCKSIZE;
+	s16 csize_n = chunksize * MAP_BLOCKSIZE;
 	// Minp/maxp of central chunk, in nodes
 	s16 ccmin = ccoff_b * MAP_BLOCKSIZE;
 	s16 ccmax = ccmin + csize_n - 1;
 	// Fullminp/fullmaxp of central chunk, in nodes
 	s16 ccfmin = ccmin - MAP_BLOCKSIZE;
 	s16 ccfmax = ccmax + MAP_BLOCKSIZE;
-	// Effective mapgen limit, in blocks
+
+	s16 mapgen_edge_min;
+	s16 mapgen_edge_max;
+
+	// Calculate SAO position limits
+	calcMapgenEdges(MAX_MAP_GENERATION_LIMIT,
+		csize_n, ccmin, ccmax, ccfmin, ccfmax, &mapgen_edge_min, &mapgen_edge_max);
+	// SAO position limits, in Irrlicht units
+	// +-3.0 is to allow SAOs closer to edges than node centre
+	m_sao_limit_min = mapgen_edge_min * BS - 3.0f;
+	m_sao_limit_max = mapgen_edge_max * BS + 3.0f;
+
+	// Calculate spawn range maximum
+	calcMapgenEdges(rangelim(mapgen_limit, 0, MAX_MAP_GENERATION_LIMIT),
+		csize_n, ccmin, ccmax, ccfmin, ccfmax, &mapgen_edge_min, &mapgen_edge_max);
```

Comment:
What's the purpose of this function? `mapgen_edge_min` and `mapgen_edge_max` do not get any new value here.

Reply:
They do, as they are calculated from the 'mapgen limit' setting instead of 'max map generation limit'.